hibernate-core/src/test/java/org/hibernate/test/type/LobUnfetchedPropertyTest.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.test.type;
+
+import static org.hibernate.testing.transaction.TransactionUtil.doInHibernate;
+
+import java.sql.Blob;
+import java.sql.Clob;
+import java.sql.NClob;
+
+import javax.persistence.Basic;
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.FetchType;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.persistence.Lob;
+
+import org.hibernate.annotations.Cache;
+import org.hibernate.annotations.CacheConcurrencyStrategy;
+import org.hibernate.testing.DialectChecks;
+import org.hibernate.testing.RequiresDialectFeature;
+import org.hibernate.testing.TestForIssue;
+import org.hibernate.testing.bytecode.enhancement.BytecodeEnhancerRunner;
+import org.hibernate.testing.junit4.BaseCoreFunctionalTestCase;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@TestForIssue(jiraKey = "HHH-12555")
+@RequiresDialectFeature(DialectChecks.SupportsExpectedLobUsagePattern.class)
+@RunWith(BytecodeEnhancerRunner.class)
+public class LobUnfetchedPropertyTest extends BaseCoreFunctionalTestCase {
+
+	@Override
+	protected Class<?>[] getAnnotatedClasses() {
+		return new Class<?>[]{ FileBlob.class, FileClob.class, FileNClob.class };
+	}
+
+	@Test
+	public void testBlob() {
+		doInHibernate( this::sessionFactory, s -> {
+			FileBlob file = new FileBlob();
+			file.setBlob( s.getLobHelper().createBlob( "TEST CASE".getBytes() ) );
+			s.save( file );
+			s.clear();
```

Comment:
If you clear the session before flushing, the entity will still be transient.

I was surprised to see that this issue arises with a transient entity.